From 9d6a9e155652884b942e750bad0119a072552db5 Mon Sep 17 00:00:00 2001
From: Mike Lothian <mike@fireburn.co.uk>
Date: Mon, 15 Dec 2025 13:40:05 +0000
Subject: [PATCH] drm/amd/display: Add HDMI 2.1 FRL and HDCP 2.3 Definitions

Introduces support structures and constants required for HDMI 2.1
Fixed Rate Link (FRL) and HDCP 2.3.

1. Updates hdcp_msg_types.h to include HDCP_VERSION_23.
2. Updates link_encoder.h to add the function pointer `enable_frl_output`.
   This allows the resource manager to dispatch FRL enablement calls
   to the appropriate DCE/DCN hardware implementation.

Signed-off-by: Mike Lothian <mike@fireburn.co.uk>
---
 drivers/gpu/drm/amd/display/dc/inc/hw/link_encoder.h | 4 ++++
 drivers/gpu/drm/amd/display/include/hdcp_msg_types.h | 1 +
 2 files changed, 5 insertions(+)

diff --git a/drivers/gpu/drm/amd/display/dc/inc/hw/link_encoder.h b/drivers/gpu/drm/amd/display/dc/inc/hw/link_encoder.h
index cff705c49564..c738f118807d 100644
--- a/drivers/gpu/drm/amd/display/dc/inc/hw/link_encoder.h
+++ b/drivers/gpu/drm/amd/display/dc/inc/hw/link_encoder.h
@@ -66,6 +66,7 @@ struct encoder_feature_support {
 			uint32_t IS_UHBR13_5_CAPABLE:1;
 			uint32_t IS_UHBR20_CAPABLE:1;
 			uint32_t DP_IS_USB_C:1;
+			uint32_t HDMI_FRL_CAPABLE:1;
 		} bits;
 		uint32_t raw;
 	} flags;
@@ -130,6 +131,9 @@ struct link_encoder_funcs {
 	void (*enable_lvds_output)(struct link_encoder *enc,
 		enum clock_source_id clock_source,
 		uint32_t pixel_clock);
+	void (*enable_frl_output)(struct link_encoder *enc,
+		const struct dc_link_settings *link_settings,
+		enum clock_source_id clock_source);
 	void (*disable_output)(struct link_encoder *link_enc,
 		enum signal_type signal);
 	void (*dp_set_lane_settings)(struct link_encoder *enc,
diff --git a/drivers/gpu/drm/amd/display/include/hdcp_msg_types.h b/drivers/gpu/drm/amd/display/include/hdcp_msg_types.h
index eced9ad91f1d..37271dca2714 100644
--- a/drivers/gpu/drm/amd/display/include/hdcp_msg_types.h
+++ b/drivers/gpu/drm/amd/display/include/hdcp_msg_types.h
@@ -81,6 +81,7 @@ enum hdcp_version {
 	HDCP_Unknown = 0,
 	HDCP_VERSION_14,
 	HDCP_VERSION_22,
+	HDCP_VERSION_23,
 };
 
 enum hdcp_link {
-- 
2.52.0

